File: python/src/common/list.py
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def list_create(list: List) -> Optional[ListNode]:
    head = tail = ListNode()
    for val in list:
        tail = list_insert(tail, val)
    return head.next


def list_insert(node: Optional[ListNode], val) -> Optional[ListNode]:
    new_node = ListNode(val, node.next)
    node.next = new_node
    return new_node


def list_equal(lhs: Optional[ListNode], rhs: Optional[ListNode]) -> bool:
    while lhs and rhs and lhs.val == rhs.val:
        lhs = lhs.next
        rhs = rhs.next
    return lhs == rhs

File: cpp/src/common/list.cpp
#include "list.h"

ListNode* list_create(std::initializer_list<int> list) {
  ListNode head;
  ListNode* tail = &head;
  for (auto val : list) {
    tail = list_insert(tail, val);
  }
  return head.next;
}

ListNode* list_insert(ListNode* node, int val) {
  ListNode* new_node = new ListNode(val, node->next);
  node->next = new_node;
  return new_node;
}

bool list_equal(const ListNode* lhs, const ListNode* rhs) {
  while (lhs != nullptr && rhs != nullptr && lhs->val == rhs->val) {
    lhs = lhs->next;
    rhs = rhs->next;
  }
  return lhs == rhs;
}